app/services/ui_notifications/clean_expired.rb

```diff
@@ -0,0 +1,41 @@
+module UINotifications
+  class CleanExpired
+    def initialize(options = {})
+      if options[:blueprint].present?
+        blueprint = NotificationBlueprint.find_by(name: options[:blueprint])
+        raise(Foreman::Exception, 'blueprint name does not exist') if blueprint.nil?
+        @blueprint = blueprint.name
+      end
+      parsed_time = options[:expired_at].is_a?(String) ? options[:expired_at].to_time : Time.now.utc
+      raise(Foreman::Exception, 'Expired time cannot be greater then now') if parsed_time > Time.now.utc
+      @expired_at = parsed_time
+      @group = options[:group]
+    end
+
+    def clean!
+      UINotifications::Base.logger.info("Removing all expired notifications")
+      to_delete = Notification.joins(:notification_blueprint)
+                      .where('notifications.expired_at < ?', @expired_at)
+                      .where(blueprint_query)
+                      .where(group_query)
+      NotificationRecipient.where(notification_id: to_delete.map {|n| n.id }).delete_all
+      @deleted_count = to_delete.delete_all
+      self
+    end
+
+    def deleted_count
+      raise 'cleaner has not cleaned anything yet, run #clean! first' if @deleted_count.nil?
+      @deleted_count
+    end
+
+    private
+
+    def blueprint_query
+      @blueprint.present? ? {:notification_blueprints => {:name => @blueprint }} : nil
```

Comment:
You could just set the instance variable in the initializer to the query instead of to the user input, and reduce these functions. (also true for the group)